primary/src/certificate_waiter.rs

```diff
@@ -158,14 +165,43 @@ impl<PublicKey: VerifyingKey> CertificateWaiter<PublicKey> {
                     }
 
                 }
+                () = &mut timer => {
+                    // We still would like to GC even if we do not receive anything from either the synchronizer or
+                    //  the Waiter. This can happen, as we may have asked for certificates that (at the time of ask)
+                    // were not past our GC bound, but which round swept up past our GC bound as we advanced rounds
+                    // & caught up with the network.
+                    // Those certificates may get stuck in pending, unless we periodically clean up.
+                    let round = self.consensus_round.load(Ordering::Relaxed);
+                    let gc_depth = self.gc_depth;
+                    if let Some(fresh_pending) = Self::clean_pending(round, gc_depth, &mut self.pending) {
+                        self.pending = fresh_pending;
+                    }
+
+                    // Reschedule the timer.
+                    timer.as_mut().reset(Instant::now() + Duration::from_millis(GC_RESOLUTION));
+                },
             }
 
-            // Cleanup internal state. Deliver the certificates waiting on garbage collected ancestors.
             let round = self.consensus_round.load(Ordering::Relaxed);
-            if round > self.gc_depth {
-                let gc_round = round - self.gc_depth;
-                self.pending = self
-                    .pending
+            let gc_depth = self.gc_depth;
+            if let Some(fresh_pending) = Self::clean_pending(round, gc_depth, &mut self.pending) {
+                self.pending = fresh_pending;
+            }
+
+            self.update_metrics();
+        }
+    }
+
+    fn clean_pending(
+        round: Round,
+        gc_depth: u64,
+        pending: &mut HashMap<HeaderDigest, (Round, oneshot::Sender<()>)>,
+    ) -> Option<HashMap<HeaderDigest, (Round, oneshot::Sender<()>)>> {
+        // Cleanup internal state. Deliver the certificates waiting on garbage collected ancestors.
+        if round > gc_depth {
+            let gc_round = round - gc_depth;
+            Some(
+                pending
```

Comment:
Since the `pending` is passed as a &mut , once we drain, isn't this going to drain the map in any case? I mean shouldn't we re-assign whatever left to the the pending list?